[M68k] always use movem for register spills

TechnoElf opened the pull request:
Fixes #106206 and #106209.

TechnoElf commented:
ping @mshockwave @0x59616e

0x59616e commented:
I apologize for the long wait. I have some free time recently and I will give some feedback ASAP.

0x59616e commented:
Thanks for this fix! I have one question: How can we ensure that our test targets the exact issue we're trying to address?

To be more specific, we are addressing two issues: spill instruction & spill size, and I don't know how these two issues are being targeted with the [llvm/test/CodeGen/M68k/PR57660.ll](https://github.com/llvm/llvm-project/pull/106715/files#diff-b49525188447bfeaedab9d052d669840b4d5edce05824f657175dba5c65003c0).

It would be great if you could give more details or add a new test. Thanks!

TechnoElf commented:
I've added some more test cases now. Do you think these adequately cover the issues addressed here?

0x59616e commented:
> To be honest, I have no idea how to test for the correct spill size. 

How about this one: 
```
$ llc ./test.ll -O0 -regalloc=fast -o - -march=m68k -stop-after=regallocfast
```

We can tell from this snippet that there is a spill slot of 2 bytes:
```
stack:
  - { id: 0, name: prev_state, type: default, offset: 0, size: 1, alignment: 1,
      stack-id: default, callee-saved-register: '', callee-saved-restored: true,
      debug-info-variable: '', debug-info-expression: '', debug-info-location: '' }
  - { id: 1, name: '', type: spill-slot, offset: 0, size: 2, alignment: 2,
      stack-id: default, callee-saved-register: '', callee-saved-restored: true,
      debug-info-variable: '', debug-info-expression: '', debug-info-location: '' }

```

Also, we can also check from this snippet that the spill instruction is movem, because it stores to the spill slot (which is %stack.1)  
```
    dead $bd1 = SUB8dd $bd1, killed $bd2, implicit-def $ccr
    MOVM16pm_P 0, %stack.1, killed $bd0 :: (store (s16) into %stack.1)
    Bne8 %bb.3, implicit killed $ccr

```

mshockwave commented on llvm/test/CodeGen/M68k/register-spills.ll:
```diff
@@ -0,0 +1,499 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -mtriple=m68k -O0 %s -o - | FileCheck %s
+
+define zeroext i1 @get1() {
```
could you split these three `get` functions and the code they're testing into a separate patch? they're not directly related to the register spilling issues you're fixing here.

TechnoElf replied:
They do test the changes made in this patch though, no? The assembly demonstrates movem being used to spill registers and 8-bit registers being assigned to 16-bit slots on the stack.

mshockwave replied:
I meant `get1`, `get8`, `get16`, and `get32`. I don't think they use movem.

TechnoElf replied:
The test requires these functions to generate values that can't be optimised away, thus forcing a register spill.
Is there a better way to create such values? I've replaced the function `define`s with `declare`s for now as the function body is unnecessary, like you said.

glaubitz commented:
@mshockwave Any chance this and #111145 could be landed?

mshockwave reviewed (approved):
LGTM w/ minor comments

TechnoElf commented:
Thanks for the review! I've implemented your requested changes

jrtc27 commented:
The commit message should have the "why" beyond linking to issue numbers. From my own research the "why" appears to be because MOVE will set condition codes, but MOVEM will not. It also appears you're conflating two changes here: one involving using MOVEM over MOVE, and one involving not promoting registers and overflowing the stack space allocated to them.

knickish commented:
@TechnoElf are you able to address the above feedback? Looks like it should just be documentation changes

TechnoElf committed:
[M68k]: do not promote moves for 8-bit and 16-bit register spills to 32-bit

8-bit and 16-bit register spills were previously always promoted to 32-bit, despite only
being assigned 16-bit stack slots. This resulted in unrelated data on the stack being
overwritten.

Fixes #106209.

TechnoElf committed:
[M68k]: always use the movem instruction for register spills

In the previous implementation, the mov instruction was used for 8-bit
register spills. However, this instruction has the side effect of
overwriting the CCR. When a spill is inserted between a compare and a
branch instruction, this would cause the necessary condition flags to be
overwritten. As 8-bit spills are already assigned 16-bit stack slots,
the 16-bit movem instruction, which doesn't affect the CCR, may be used
instead.

Fixes #106209.

TechnoElf commented:
@jrtc27 @mshockwave @knickish I've split up the original commit and added more details to the new commit messages.

knickish commented:
Trying to repro this, may be spurious as this build doesn't even appear to target m68k

```
...
-- Targeting LoongArch
-- Targeting Mips
-- Targeting MSP430
-- Targeting NVPTX
...
```

s-barannikov commented:
Sanitizer tests fail from time to time, you can ignore this failure.
